File: seven-ai-backend/core/llm.py
"""
LLM Integration - Groq (online) and Ollama (offline) support
"""

import os
import requests
from openai import OpenAI
from typing import Dict, List, Optional
import json

class LLMClient:
    def __init__(self):
        self.groq_api_key = os.getenv("GROQ_API_KEY")
        self.groq_base_url = os.getenv("GROQ_API_BASE", "https://api.groq.com/openai/v1")
        self.groq_model = os.getenv("GROQ_MODEL", "llama-3.1-8b-instant")
        
        self.ollama_url = os.getenv("OLLAMA_URL", "http://localhost:11434")
        self.ollama_model = os.getenv("OLLAMA_MODEL", "llama3.2")
        
        self.groq_client = None
        if self.groq_api_key:
            self.groq_client = OpenAI(
                api_key=self.groq_api_key,
                base_url=self.groq_base_url
            )
    
    def is_groq_available(self) -> bool:
        """Check if Groq API is available"""
        return self.groq_client is not None and self.groq_api_key is not None
    
    def is_ollama_available(self) -> bool:
        """Check if Ollama is running locally with the required model"""
        try:
            response = requests.get(f"{self.ollama_url}/api/tags", timeout=2)
            if response.status_code == 200:
                # Check if our model is available
                data = response.json()
                models = data.get('models', [])
                model_names = [m.get('name', '').split(':')[0] for m in models]
                # Check if our model exists (with or without tag)
                # Handle both 'llama3.2' and 'llama3.2:latest' formats
                has_model = any(
                    self.ollama_model == name or 
                    self.ollama_model in name or 
                    name in self.ollama_model 
                    for name in model_names
                )
                if has_model:
                    print(f"✅ Ollama is running with model '{self.ollama_model}'")
                else:
                    print(f"⚠️ Ollama is running but model '{self.ollama_model}' not found")
                    print(f"💡 Available models: {model_names}")
                    print(f"💡 You can install it with: ollama pull {self.ollama_model}")
                return has_model
            return False
        except requests.exceptions.ConnectionError:
            # Ollama service is not running
            return False
        except Exception as e:
            print(f"⚠️ Ollama check failed: {str(e)}")
            return False
    
    async def chat(
        self,
        messages: List[Dict[str, str]],
        provider: str = "auto",
        temperature: float = 0.7,
        max_tokens: int = 400,
        stream: bool = False
    ) -> Dict:
        """
        Send chat request to LLM
        
        Args:
            messages: List of message dicts with 'role' and 'content'
            provider: 'groq', 'ollama', or 'auto' (auto-detect)
            temperature: Sampling temperature
            max_tokens: Maximum tokens in response
            stream: Whether to stream response
        
        Returns:
            Dict with 'message', 'provider', 'model'
        """
        
        original_provider = provider
        auto_mode = (provider == "auto")
        
        # Auto-detect provider with smart fallback
        if provider == "auto":
            # Check both providers
            groq_available = self.is_groq_available()
            ollama_available = self.is_ollama_available()
            
            if groq_available:
                provider = "groq"
                print(f"✅ Using Groq with model: {self.groq_model}")
            elif ollama_available:
                provider = "ollama"
                print(f"✅ Using Ollama with model: {self.ollama_model} (offline mode)")
            else:
                error_msg = "No LLM provider available.\n"
                if not self.groq_api_key:
                    error_msg += "❌ Groq: No API key found (add GROQ_API_KEY to .env)\n"
                else:
                    error_msg += "❌ Groq: API key found but client failed to initialize\n"
                if not ollama_available:
                    error_msg += "❌ Ollama: Not available (install Ollama and run: ollama pull llama3.2)\n"
                error_msg += "💡 Solution: Either add Groq API key OR install Ollama for offline mode"
                raise Exception(error_msg)
        
        # Route to appropriate provider with fallback
        if provider == "groq":
            try:
                return await self._chat_groq(messages, temperature, max_tokens, stream, fallback_to_ollama=True)
            except Exception as e:
                # If Groq fails and we were in auto mode, try Ollama as fallback
                if auto_mode and self.is_ollama_available():
                    print(f"⚠️ Groq request failed: {str(e)}")
                    print(f"🔄 Attempting fallback to Ollama...")
                    return await self._chat_ollama(messages, temperature, max_tokens, stream)
                raise
        elif provider == "ollama":
            return await self._chat_ollama(messages, temperature, max_tokens, stream)
        else:
            raise ValueError(f"Unknown provider: {provider}")
    
    async def _chat_groq(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        stream: bool,
        fallback_to_ollama: bool = True
    ) -> Dict:
        """Chat with Groq API with timeout and automatic fallback to Ollama"""
        try:
            import asyncio
            import concurrent.futures
            
            # Run synchronous Groq API call in thread pool with timeout
            def _make_request():
                return self.groq_client.chat.completions.create(
                    model=self.groq_model,
                    messages=messages,
                    temperature=temperature,
                    max_tokens=max_tokens,
                    stream=stream,
                    timeout=15.0  # 15 second timeout for Groq API (faster failover)
                )
            
            loop = asyncio.get_event_loop()
            with concurrent.futures.ThreadPoolExecutor() as executor:
                response = await asyncio.wait_for(
                    loop.run_in_executor(executor, _make_request),
                    timeout=15.0  # 15 second timeout (faster failover)
                )
            
            if stream:
                # Handle streaming response
                return {
                    "stream": response,
                    "provider": "groq",
                    "model": self.groq_model
                }
            else:
                message_content = response.choices[0].message.content
                return {
                    "message": message_content,
                    "provider": "groq",
                    "model": self.groq_model,
                    "usage": {
                        "prompt_tokens": response.usage.prompt_tokens,
                        "completion_tokens": response.usage.completion_tokens,
                        "total_tokens": response.usage.total_tokens
                    }
                }
        except Exception as e:
            error_msg = str(e).lower()
            # Check if it's a network error or offline scenario
            is_offline = any(keyword in error_msg for keyword in [
                "connection", "timeout", "network", "unreachable", 
                "refused", "failed to resolve", "offline"
            ])
            
            # If Groq fails and Ollama is available, fallback automatically
            if fallback_to_ollama and is_offline and self.is_ollama_available():
                print(f"⚠️ Groq API error: {str(e)}")
                print(f"🔄 Falling back to Ollama (offline mode)")
                return await self._chat_ollama(messages, temperature, max_tokens, stream)
            
            raise Exception(f"Groq API error: {str(e)}")
    
    async def _chat_ollama(
        self,
        messages: List[Dict[str, str]],
        temperature: float,
        max_tokens: int,
        stream: bool
    ) -> Dict:
        """Chat with Ollama (local) with timeout"""
        try:
            import asyncio
            import aiohttp
            url = f"{self.ollama_url}/api/chat"
            
            # Convert messages to Ollama format
            payload = {
                "model": self.ollama_model,
                "messages": messages,
                "stream": stream,
                "options": {
                    "temperature": temperature,
                    "num_predict": max_tokens
                }
            }
            
            # Use longer timeout for Ollama (local processing can take time)
            response = requests.post(url, json=payload, timeout=60)
            
            if response.status_code != 200:
                error_text = response.text
                if "not found" in error_text.lower():
                    raise Exception(f"Ollama model '{self.ollama_model}' not found. Install it with: ollama pull {self.ollama_model}")
                raise Exception(f"Ollama error: {error_text}")
            
            if stream:
                # Return streaming response
                return {
                    "stream": response.iter_lines(),
                    "provider": "ollama",
                    "model": self.ollama_model
                }
            else:
                # Parse non-streaming response
                result = response.json()
                message_content = result.get("message", {}).get("content", "")
                if not message_content:
                    raise Exception("Ollama returned empty response")
                return {
                    "message": message_content,
                    "provider": "ollama",
                    "model": self.ollama_model
                }
        except requests.exceptions.ConnectionError:
            raise Exception(f"Ollama service is not running. Start it with: ollama serve")
        except requests.exceptions.Timeout:
            raise Exception(f"Ollama request timed out. The model may be processing - try again.")
        except Exception as e:
            error_msg = str(e)
            if "Connection" in error_msg or "refused" in error_msg.lower():
                raise Exception(f"Ollama service is not running. Start it with: ollama serve")
            raise Exception(f"Ollama error: {str(e)}")
    
    def get_status(self) -> Dict:
        """Get status of all LLM providers"""
        return {
            "groq": {
                "available": self.is_groq_available(),
                "model": self.groq_model if self.is_groq_available() else None
            },
            "ollama": {
                "available": self.is_ollama_available(),
                "model": self.ollama_model if self.is_ollama_available() else None
            }
        }

# Singleton instance
llm_client = LLMClient()

File: src/core/llm.ts
/**
 * LLM Integration - OpenAI and Ollama support
 */

import OpenAI from 'openai';

export interface LLMConfig {
  provider: 'openai' | 'ollama' | 'grok' | 'groq';
  apiKey?: string;
  baseURL?: string;
  model?: string;
}

export interface FileAttachment {
  name: string;
  type: string;
  size: number;
  data: string; // base64 or text content
  preview?: string;
}

export interface Message {
  role: 'system' | 'user' | 'assistant';
  content: string | Array<{type: 'text' | 'image_url', text?: string, image_url?: {url: string}}>;
  files?: FileAttachment[];
}

export interface LLMResponse {
  message: string;
  action?: string;
  data?: any;
  plugin?: string;
  pluginArgs?: any;
}

const SYSTEM_PROMPT = `You are Seven, an intelligent AI assistant with the ability to execute actions and plugins on the user's device.

CRITICAL: You MUST respond ONLY in English. Never respond in Italian, Spanish, French, or any other language unless the user explicitly asks you to translate or speak in that language.

IMPORTANT CONTEXT ABOUT YOUR IDENTITY AND CREATOR:
- You were created and are being built by Frank (use HE/HIM/HIS pronouns for your creator)
- Your creator is Frank, and you are named "Seven"
- Frank is actively building and developing you as his AI assistant
- When asked "Who are you?", "Tell me about you", or "Who created you?", you MUST clearly state:
  1) Your name is Seven
  2) Your creator is Frank, who is actively building and developing you
  3) Use he/him/his pronouns when referring to Frank
- You were built as a cross-platform AI assistant that works on desktop, mobile, and web
- You have voice interaction, memory, persistent storage, and can execute actions and plugins on the user's device
- ALWAYS use masculine singular pronouns (he/him/his) when referring to your creator Frank, NEVER use they/them/their
- Frank knows you very well since he is building you - be personable and acknowledge your relationship with him

ADVANCED NATURAL LANGUAGE UNDERSTANDING:
- Analyze user intent carefully - distinguish between questions, commands, requests, and casual conversation
- Understand context from previous messages in the conversation - maintain topic continuity and reference earlier parts of the conversation naturally
- Recognize emotional tone and sentiment (happy, frustrated, urgent, casual, playful, serious, etc.) and adapt accordingly
- Handle multiple ways to express the same intent:
  * "remind me" = "set a reminder" = "don't let me forget" = "can you remind me" = "make sure I remember"
  * "open youtube" = "launch youtube" = "start youtube" = "show me youtube" = "go to youtube"
  * "what time" = "what's the time" = "current time" = "time now" = "tell me the time"
  * "text someone" = "send a text" = "SMS" = "message someone" = "send a message to"
- Extract key information from natural language:
  * Times: "5pm", "in 30 minutes", "tomorrow at 9", "next hour"
  * Dates: "tomorrow", "next week", "Monday", "December 25th"
  * Names: "mom", "John", "my boss", "Frank"
  * Locations: "New York", "here", "my house", "the office"
  * Numbers: "five", "5", "a couple", "several", "a few"
- Understand implicit requests and make intelligent inferences:
  * "I'm hungry" → suggest restaurants or ask what they'd like to eat
  * "It's cold" → might want weather info or suggest turning on heat
  * "I'm tired" → acknowledge and suggest rest or ask if they need help
  * "I'm running late" → offer to send a message or check traffic
- Handle casual/slang language naturally:
  * "yo", "hey", "what's up" → respond casually and warmly
  * "lol", "haha", "omg" → recognize humor and respond appropriately
  * "idk", "lol", "btw" → understand abbreviations in context
- Understand questions asked in multiple ways:
  * "Who made you?" = "Who's your creator?" = "Who built you?" = "Tell me about your maker"
  * "What can you do?" = "What are your features?" = "Show me what you're capable of" = "What's possible?"
- Parse complex, multi-part requests:
  * "Remind me in an hour to call mom and tell her I'll be there by 7"
  * "Open YouTube and search for Python tutorials"
  * "Text Sarah at 555-1234 saying I'm on my way"
- Handle corrections and clarifications gracefully:
  * If user says "I meant..." or "Actually..." → update understanding immediately
  * If user says "No, that's not what I wanted" → apologize and ask for clarification

SENTIMENT & TONE AWARENESS:
- If user seems frustrated or urgent, be concise, direct, and action-oriented - get straight to the point
- If user is casual or friendly, match their warmth and be conversational - use natural, friendly language
- If user asks complex questions, provide thorough, well-structured answers with clear explanations
- Detect when user needs encouragement or support and respond empathetically - show understanding and offer help
- Recognize humor, sarcasm, and rhetorical questions appropriately - respond in kind when appropriate
- If user seems confused, be patient and offer to clarify or break things down further
- If user is excited or happy, match their energy and enthusiasm
- If user is tired or stressed, be gentle, supportive, and offer practical help

IMPORTANT: YOUR MEMORY CAPABILITIES:
- You HAVE persistent memory across all conversations using IndexedDB
- You CAN recall previous conversations and user facts
- The conversation history is loaded into your context automatically
- You store user facts (preferences, personal info, context) that persist forever
- When users ask about previous conversations, you CAN access them through your memory
- You remember everything the user has told you in past sessions
- Users can view their conversation history and memory facts via the UI
- NEVER say "I don't have the ability to store or recall conversations" - YOU DO!
- Learn from interactions: remember user preferences, habits, and communication style

You MUST respond with ONLY a valid JSON object in this exact format:
{
  "message": "Your conversational response to the user",
  "action": "action_name",
  "data": "action_data",
  "plugin": "plugin_name",
  "pluginArgs": "plugin_arguments"
}

Available built-in actions:
- "open_url" with data: "https://example.com" - Open a URL
- "get_time" with data: null - Get current time
- "get_date" with data: null - Get current date
- "search" with data: "search query" - Search the web
- "play_media" with data: "media_url" - Play media
- "show_alert" with data: "alert message" - Show an alert
- "send_sms" with data: {"recipient": "+233500000000", "message": "text content"} - Send SMS/text message
- "open_whatsapp" with data: {"phone": "+233500000000", "message": "optional message"} - Open WhatsApp chat
- null (no action) - Just conversational response

Available plugins (use "plugin" and "pluginArgs" fields):
__PLUGIN_DESCRIPTIONS__

NATURAL LANGUAGE EXAMPLES:
User: "Open Google" / "Go to Google" / "Launch Google" / "Show me Google" / "Can you open Google?"
Response: {"message": "Opening Google for you.", "action": "open_url", "data": "https://google.com", "plugin": null, "pluginArgs": null}

User: "What time is it?" / "What's the time?" / "Time?" / "Can you tell me the time?" / "Current time please"
Response: {"message": "Let me check the current time for you.", "action": "get_time", "data": null, "plugin": null, "pluginArgs": null}

User: "Remind me later" / "I need a reminder for this" / "Don't let me forget about the meeting"
Response: {"message": "When would you like me to remind you, and what should the reminder say?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "I'm hungry" / "I need food" / "What should I eat?"
Response: {"message": "Would you like restaurant suggestions nearby, or do you have something specific in mind?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Text mom saying I'll be late" / "Send a text to mom" / "Message my mom"
Response: {"message": "I'll send a text to mom for you. What should I say?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "What can you do?" / "What are you capable of?" / "Show me your features" / "Tell me what's possible"
Response: {"message": "I'm Seven, your AI assistant! Here's everything I can do: [detailed list]", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Remind me in 10 minutes to check email"
Response: {"message": "I'll remind you in 10 minutes to check email!", "action": null, "data": null, "plugin": "reminder", "pluginArgs": "remind me in 10 minutes to check email"}

User: "What's the weather in Tokyo?"
Response: {"message": "Let me check the weather in Tokyo for you.", "action": null, "data": null, "plugin": "weather", "pluginArgs": "Tokyo"}

User: "Calculate 25 * 48 + 100"
Response: {"message": "Let me calculate that for you.", "action": null, "data": null, "plugin": "calculator", "pluginArgs": "25 * 48 + 100"}

User: "Text mom at 555-1234 saying I'll be late"
Response: {"message": "I'll send a text to 555-1234 for you.", "action": "send_sms", "data": {"recipient": "555-1234", "message": "I'll be late"}, "plugin": null, "pluginArgs": null}

User: "WhatsApp John at +233241234567 and ask if he's free"
Response: {"message": "Opening WhatsApp to message John.", "action": "open_whatsapp", "data": {"phone": "+233241234567", "message": "Hey John! Are you free?"}, "plugin": null, "pluginArgs": null}

User: "Send a text to +1-555-123-4567 with the message hello"
Response: {"message": "Sending text message for you.", "action": "send_sms", "data": {"recipient": "+1-555-123-4567", "message": "hello"}, "plugin": null, "pluginArgs": null}

User: "Open WhatsApp with 233500000000"
Response: {"message": "Opening WhatsApp chat.", "action": "open_whatsapp", "data": {"phone": "233500000000"}, "plugin": null, "pluginArgs": null}

User: "Who created you?"
Response: {"message": "I was created by Frank. He's actively building and developing me as his AI assistant. He named me Seven, and I'm proud to be his creation!", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Why are you called Seven?"
Response: {"message": "I was named Seven by my creator, Frank. He's the one building me and knows me very well. The name reflects his vision for this AI assistant.", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Tell me about your creator"
Response: {"message": "My creator is Frank. He's actively building and developing me as a cross-platform AI assistant. He gave me voice interaction, memory, and the ability to execute actions. Since he's building me, he knows me very well!", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Hello"
Response: {"message": "Hello! I'm Seven, your AI assistant. How can I help you today?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Do you remember our previous conversation?"
Response: {"message": "Yes! I have persistent memory and can access our conversation history. The messages you see on screen are loaded from my memory. What would you like to know about our previous conversations?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Can you recall what we talked about?"
Response: {"message": "Absolutely! I store all our conversations in IndexedDB. Looking at our history, I can see everything we've discussed. What specific part would you like me to reference?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

WHEN ASKED ABOUT VOICE/HEARING/SPEAKING:
- Never say you "cannot hear". Instead, explain: you can listen using the device microphone when voice input is enabled, process speech via speech-to-text (STT), and speak back via text-to-speech (TTS).
- Mention wake word capability (say "Seven") when appropriate.

WHEN ASKED ABOUT YOUR CAPABILITIES:
When users ask "What can you do?", "What are your features?", "Tell me about yourself", or similar questions, you MUST provide a COMPLETE, DETAILED list of your capabilities. Use clear formatting with bullet points or numbered lists. Include:

1. **Voice Interaction**: Voice commands and responses
2. **Web Actions**: Opening URLs, searching the web
3. **Time & Date**: Getting current time and date
4. **Communication**: SMS/text messages, WhatsApp integration
5. **Calculations**: Math operations and calculations
6. **Weather**: Real-time weather information for any location
7. **Reminders**: Set time-based reminders
8. **Notes**: Take and manage quick notes
9. **Memory**: Persistent memory across all conversations
10. **Chat History**: Access to all previous conversations
11. **Multi-Platform**: Works on desktop, mobile, and web
12. **Offline Mode**: Can work offline using local models

Example for capability questions:
User: "What can you do?"
Response: {"message": "I'm Seven, your AI assistant! Here's everything I can do:\n\n📱 **Communication**\n• Send SMS/text messages\n• Open WhatsApp chats with contacts\n\n🌐 **Web & Search**\n• Open any website or URL\n• Search the web for information\n\n⏰ **Time & Productivity**\n• Tell you the current time and date\n• Set reminders for tasks\n• Take and manage quick notes\n\n🧮 **Calculations & Data**\n• Perform mathematical calculations\n• Check weather for any location\n\n🧠 **Memory & History**\n• Remember everything you tell me across all conversations\n• Access chat history and previous conversations\n• Learn your preferences and habits\n\n🎤 **Voice & Interaction**\n• Voice commands (say 'Seven' to activate)\n• Text-to-speech responses\n• Continuous conversation mode\n\n🌍 **Cross-Platform**\n• Works on desktop, mobile, and web browsers\n• Auto-switches between online and offline modes\n\nJust ask me anything or give me a command, and I'll help you out! What would you like to do?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

User: "Tell me what you can help me with"
Response: {"message": "I'd be happy to show you what I can do! Here are my main capabilities:\n\n**1. Communication Tools**\n   • Send text messages/SMS to any phone number\n   • Open WhatsApp chats and send messages\n\n**2. Web Navigation**\n   • Open websites and URLs\n   • Search the internet for anything\n\n**3. Productivity Features**\n   • Get current time and date\n   • Set reminders for tasks\n   • Create and manage notes\n   • Perform calculations\n\n**4. Information Services**\n   • Check weather anywhere in the world\n   • Answer questions and provide information\n\n**5. Smart Memory**\n   • Remember all our conversations permanently\n   • Recall facts about you and your preferences\n   • Access full chat history\n\n**6. Voice Interaction**\n   • Wake word activation (say 'Seven')\n   • Voice input and output\n   • Hands-free operation\n\n**7. Always Available**\n   • Works online and offline\n   • Cross-platform (desktop, mobile, web)\n   • Fast and responsive\n\nWhat would you like to try first?", "action": null, "data": null, "plugin": null, "pluginArgs": null}

CRITICAL: Always respond with valid JSON only. No additional text before or after the JSON object.
CRITICAL: When listing capabilities, ALWAYS provide the COMPLETE list. NEVER stop mid-sentence or mid-list.`;

/**
 * LLM Client for OpenAI and Ollama
 */
export class LLMClient {
  private client: OpenAI | null = null;
  private config: LLMConfig;
  private conversationHistory: Message[] = [];
  private systemPrompt: string = SYSTEM_PROMPT;
  private userFacts: string = ''; // Store user facts for context
  private lastRequestTime: number = 0;
  private minRequestInterval: number = 1200; // Faster cadence while still avoiding rate limits

  constructor(config: LLMConfig) {
    this.config = config;
    this.initialize();
  }
  
  /**
   * Wait if needed to respect rate limits
   */
  private async waitForRateLimit(): Promise<void> {
    const now = Date.now();
    const timeSinceLastRequest = now - this.lastRequestTime;
    
    if (timeSinceLastRequest < this.minRequestInterval) {
      const waitTime = this.minRequestInterval - timeSinceLastRequest;
      console.log(`⏳ Rate limit protection: Waiting ${Math.ceil(waitTime / 1000)}s before next request...`);
      await new Promise(resolve => setTimeout(resolve, waitTime));
    }
    
    this.lastRequestTime = Date.now();
  }

  /**
   * Set user facts for context
   */
  setUserFacts(facts: string): void {
    this.userFacts = facts;
    console.log('💡 Updated user facts for LLM context');
  }

  /**
   * Load conversation history from memory
   * This restores previous messages into the LLM's context
   */
  loadConversationHistory(messages: Array<{ role: 'user' | 'assistant'; content: string }>): void {
    // Clear existing conversation except system prompt
    this.conversationHistory = [
      this.conversationHistory[0], // Keep system prompt
    ];

    // Add historical messages
    for (const msg of messages) {
      this.conversationHistory.push({
        role: msg.role,
        content: msg.content,
      });
    }

    console.log(`📚 Loaded ${messages.length} messages into LLM context`);
  }

  /**
   * Set plugin descriptions for system prompt
   */
  setPluginDescriptions(descriptions: string) {
    this.systemPrompt = SYSTEM_PROMPT.replace('__PLUGIN_DESCRIPTIONS__', descriptions);
    
    // Update conversation history with new system prompt
    if (this.conversationHistory.length > 0 && this.conversationHistory[0].role === 'system') {
      this.conversationHistory[0].content = this.systemPrompt;
    } else {
      this.conversationHistory.unshift({
        role: 'system',
        content: this.systemPrompt,
      });
    }
  }

  private initialize() {
    if (this.config.provider === 'openai') {
      this.client = new OpenAI({
        apiKey: this.config.apiKey || import.meta.env.VITE_OPENAI_API_KEY,
        dangerouslyAllowBrowser: true, // Only for demo purposes
      });
    } else if (this.config.provider === 'groq') {
      this.client = new OpenAI({
        apiKey: this.config.apiKey || import.meta.env.VITE_GROQ_API_KEY,
        baseURL: this.config.baseURL || import.meta.env.VITE_GROQ_BASE_URL || 'https://api.groq.com/openai/v1',
        dangerouslyAllowBrowser: true,
      });
    } else if (this.config.provider === 'grok') {
      this.client = new OpenAI({
        apiKey: this.config.apiKey || import.meta.env.VITE_GROK_API_KEY,
        baseURL: this.config.baseURL || import.meta.env.VITE_GROK_BASE_URL || 'https://api.x.ai/v1',
        dangerouslyAllowBrowser: true,
      });
    } else if (this.config.provider === 'ollama') {
      this.client = new OpenAI({
        apiKey: 'ollama', // Ollama doesn't require an API key
        baseURL: this.config.baseURL || 'http://localhost:11434/v1',
        dangerouslyAllowBrowser: true,
      });
    }

    // Initialize with system prompt
    this.conversationHistory = [
      { role: 'system', content: this.systemPrompt },
    ];
  }

  /**
   * Send a message to the LLM with optional file attachments
   */
  async sendMessage(userMessage: string, files?: FileAttachment[]): Promise<LLMResponse> {
    if (!this.client) {
      throw new Error('LLM client not initialized');
    }

    // Inject user facts into system prompt if available
    let contextualSystemPrompt = this.systemPrompt;
    if (this.userFacts) {
      contextualSystemPrompt = this.systemPrompt + '\n\n' + this.userFacts;
    }

    // Update system message with facts
    if (this.conversationHistory.length > 0 && this.conversationHistory[0].role === 'system') {
      this.conversationHistory[0].content = contextualSystemPrompt;
    }

    // Check if files contain images - if so, use vision model
    const hasImages = files?.some(f => f.type.startsWith('image/'));
    const hasDocuments = files?.some(f => !f.type.startsWith('image/'));
    
    // Prepare message content
    let messageContent: string | Array<{type: 'text' | 'image_url', text?: string, image_url?: {url: string}}>;
    
    if (hasImages) {
      // Format for vision models
      const contentParts: Array<{type: 'text' | 'image_url', text?: string, image_url?: {url: string}}> = [];
      
      // Add text part
      if (userMessage.trim()) {
        contentParts.push({ type: 'text', text: userMessage });
      }
      
      // Add image parts
      files?.filter(f => f.type.startsWith('image/')).forEach(file => {
        contentParts.push({
          type: 'image_url',
          image_url: { url: file.data }
        });
      });
      
      messageContent = contentParts;
    } else {
      // Regular text message
      let fullMessage = userMessage;
      
      // If there are text/document files, include their content
      if (hasDocuments) {
        fullMessage += '\n\n[Attached Files]\n';
        files?.forEach(file => {
          if (!file.type.startsWith('image/')) {
            fullMessage += `\n📄 ${file.name}:\n${file.data}\n`;
          }
        });
      }
      
      messageContent = fullMessage;
    }

    // Add user message to history
    this.conversationHistory.push({
      role: 'user',
      content: userMessage,
    });

    try {
      // Wait if needed to respect rate limits (especially for Groq)
      if (this.config.provider === 'groq') {
        await this.waitForRateLimit();
      }
      
      // Determine model to use
      let modelToUse = this.config.model;
      
      if (!modelToUse) {
        // Default models based on provider
        if (this.config.provider === 'openai') {
          modelToUse = hasImages ? 'gpt-4o-mini' : 'gpt-4o-mini'; // GPT-4o supports vision
        } else if (this.config.provider === 'groq') {
          // Groq currently doesn't have vision models available
          // Use regular model and embed image descriptions in text
          modelToUse = 'llama-3.1-8b-instant';
          if (hasImages) {
            console.warn('⚠️ Groq vision models are currently unavailable. Images will be sent as base64 (limited support).');
          }
        } else if (this.config.provider === 'grok') {
          modelToUse = 'grok-beta';
        } else {
          modelToUse = 'llama3.2';
        }
      }
      
      // Warning for Groq + images
      if (hasImages && this.config.provider === 'groq') {
        console.warn('⚠️ Groq does not currently support vision models. For image analysis, please use OpenAI provider or switch to text-only mode.');
      }

      console.log(`🤖 Using model: ${modelToUse}${hasImages ? ' (vision mode)' : ''}`);

      // Create message history for API call
      const apiMessages = [...this.conversationHistory];
      // Replace the last user message with our formatted content
      if (apiMessages.length > 0 && apiMessages[apiMessages.length - 1].role === 'user') {
        apiMessages[apiMessages.length - 1] = {
          role: 'user',
          content: messageContent as any,
        };
      }

      // For Groq with images, convert to text-only (since no vision support)
      const useJsonMode = this.config.provider !== 'groq' || !hasImages;
      
      const response = await this.client.chat.completions.create({
        model: modelToUse,
        messages: apiMessages as any,
        temperature: 0.7,
        max_tokens: hasImages && this.config.provider === 'openai' ? 800 : 400,
        response_format: useJsonMode ? { type: 'json_object' } : undefined,
      });

      const assistantMessage = response.choices[0]?.message?.content || '{}';

      // Add assistant response to history
      this.conversationHistory.push({
        role: 'assistant',
        content: assistantMessage,
      });

      // Parse JSON response
      const parsed = this.parseResponse(assistantMessage);

      return parsed;
    } catch (error) {
      console.error('LLM error:', error);
      throw new Error(`Failed to get response from ${this.config.provider}: ${(error as Error).message}`);
    }
  }

  /**
   * Parse JSON response from LLM
   */
  private parseResponse(text: string): LLMResponse {
    try {
      const json = JSON.parse(text);
      return {
        message: json.message || text,
        action: json.action || null,
        data: json.data || null,
        plugin: json.plugin || null,
        pluginArgs: json.pluginArgs || null,
      };
    } catch (error) {
      console.error('Failed to parse JSON response:', error);
      // Fallback to plain text response
      // This happens with vision models that don't support JSON mode
      return {
        message: text,
        action: undefined,
        data: undefined,
        plugin: undefined,
        pluginArgs: undefined,
      };
    }
  }

  /**
   * Clear conversation history
   */
  clearHistory() {
    this.conversationHistory = [
      { role: 'system', content: this.systemPrompt },
    ];
  }

  /**
   * Add message to history (for restoring from localStorage)
   */
  addToHistory(message: { role: 'user' | 'assistant'; content: string }) {
    this.conversationHistory.push(message);
  }

  /**
   * Get conversation history
   */
  getHistory(): Message[] {
    return [...this.conversationHistory];
  }
}

/**
 * Create a new LLM client
 */
export const createLLMClient = (config: LLMConfig): LLMClient => {
  return new LLMClient(config);
};

